File: server/src/jobs/indexingQueue.js
// Target design: in-memory FIFO queue today, swappable for BullMQ/Redis
// later without changing callers. Lets /api/upload return immediately while
// chunking+embedding happens in the background; chat polls indexing_jobs
// (see migrations/002_knowledge_metadata.sql) for status.
const queue = [];
let processing = false;

export function enqueueIndexingJob(job) {
  queue.push(job);
  processQueue();
}

async function processQueue() {
  if (processing) return;
  processing = true;
  while (queue.length > 0) {
    const job = queue.shift();
    try {
      await job.run();
    } catch (err) {
      // TODO: persist failure to indexing_jobs.status = 'failed' once that
      // table is migrated in; for now just don't crash the queue.
      console.error('indexingQueue job failed:', err);
    }
  }
  processing = false;
}

File: server/app/jobs/indexing_queue.py
import asyncio
from dataclasses import dataclass
from typing import Awaitable, Callable

from ..core.logger import logger

_queue: asyncio.Queue = asyncio.Queue()
_processing = False


@dataclass
class IndexingJob:
    run: Callable[[], Awaitable[None]]


def enqueue_indexing_job(job: IndexingJob):
    _queue.put_nowait(job)
    asyncio.create_task(_process_queue())


async def _process_queue():
    global _processing
    if _processing:
        return
    _processing = True
    try:
        while not _queue.empty():
            job = await _queue.get()
            try:
                await job.run()
            except Exception as err:
                logger.error("indexingQueue job failed", {"error": str(err)})
    finally:
        _processing = False
